### PR TITLE
Move allocator functions to their own file

### PR DESCRIPTION
**Pull request by danielsn**
*Description of changes:*
1. Move allocator functions to their own file
2. Create a stubs for the allocator functions that will be much more performant under CBMC

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/allocator.c [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    /* Since the allocator doesn't support realloc, we'll need to emulate it (inefficiently). */
+    if (oldsize >= newsize) {
+        return AWS_OP_SUCCESS;
+    }
+
+    void *newptr = allocator->mem_acquire(allocator, newsize);
+    if (!newptr) {
+        aws_raise_error(AWS_ERROR_OOM);
```
Comment (justinboswell) [inline source/allocator.c]: Can just ```return aws_raise_error(AWS_ERROR_OOM);```

File: .cbmc-batch/source/proof_allocators.c [hunk 2]
```diff
@@ -105,3 +107,131 @@ void *can_fail_realloc(void *ptr, size_t newsize) {
...
+
+    /* Defensive check: never use calloc with size * num that would overflow
+     * https://wiki.sei.cmu.edu/confluence/display/c/MEM07-C.+Ensure+that+the+arguments+to+calloc%28%29%2C+when+multiplied%2C+do+not+wrap
+     */
+    size_t required_bytes;
+    if (aws_mul_size_checked(num, size, &required_bytes)) {
+        return NULL;
+    }
+
+    /* If there is a defined calloc, use it */
```
Comment (justinboswell) [inline .cbmc-batch/source/proof_allocators.c]: There appears to always be a defined calloc :)

File: source/allocator.c (hunk 1 above)
Comment (danielsn) [reply source/allocator.c]: Done

File: .cbmc-batch/source/proof_allocators.c (hunk 2 above)
Comment (danielsn) [reply .cbmc-batch/source/proof_allocators.c]: Fixed.

**Pull request merged**